vllm/v1/worker/gpu_model_runner.py

```diff
@@ -103,6 +103,41 @@
 logger = init_logger(__name__)
 
 
+# Wrapper for ModelRunnerOutput to support overlapped execution.
+class AsyncGPUModelRunnerOutput(AsyncModelRunnerOutput):
+
+    def __init__(
+        self,
+        model_runner_output: ModelRunnerOutput,
+        sampled_token_ids: torch.Tensor,
+        invalid_req_indices: list[int],
+        async_output_copy_stream: torch.cuda.Stream,
+    ):
+        self.model_runner_output = model_runner_output
+        self.sampled_token_ids = sampled_token_ids
+        self.invalid_req_indices = invalid_req_indices
+        self.async_output_copy_stream = async_output_copy_stream
+
+    def copy_to_host(self) -> ModelRunnerOutput:
+        """Copy the device tensors to the host and return a ModelRunnerOutput.
+        
+        This function blocks until the copy is finished.
+        """
+        default_stream = torch.cuda.current_stream()
+        with torch.cuda.stream(self.async_output_copy_stream):
+            self.async_output_copy_stream.wait_stream(default_stream)
+            sampled_token_ids_cpu = self.sampled_token_ids.to(
+                'cpu', non_blocking=True)
+        self.async_output_copy_stream.synchronize()
```

Comment:
I think this additional line may be needed for (re)allocation safety
```suggestion
        self.sampled_token_ids.record_stream(self.async_output_copy_stream)
        self.async_output_copy_stream.synchronize()
```

Reply:
I did some reading here: https://docs.pytorch.org/docs/stable/generated/torch.Tensor.record_stream.html

While this would work, and would allow us to no longer hold a reference to the GPU tensor, I am concerned about the performance impact:
> These safety guarantees come at some performance and predictability cost (analogous to the tradeoff between GC and manual memory management), so if you are in a situation where you manage the full lifetime of your tensors, you may consider instead manually managing CUDA events so that calling this method is not necessary. In particular, when you call this method, on later allocations the allocator will poll the recorded stream to see if all operations have completed yet

So, I've decided to just keep the GPU tensor reference on the object. As long as a reference is held I believe the tensor's memory lifetime will be tied to that of the AsyncGPUModelRunnerOutput, which seems ideal.